File: core/src/test/java/dev/morphia/critter/parser/GeneratorTest.java
package dev.morphia.critter.parser;

import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import dev.morphia.config.MorphiaConfig;
import dev.morphia.critter.ClassfileOutput;
import dev.morphia.critter.CritterClassLoader;
import dev.morphia.critter.parser.gizmo.CritterGizmoGenerator;
import dev.morphia.critter.parser.gizmo.GizmoEntityModelGenerator;
import dev.morphia.critter.sources.Example;
import dev.morphia.mapping.ReflectiveMapper;
import dev.morphia.mapping.codec.pojo.critter.CritterEntityModel;

import io.github.classgraph.ClassGraph;

public class GeneratorTest {
    public static final CritterEntityModel entityModel;
    public static final CritterClassLoader critterClassLoader = new CritterClassLoader();

    static {
        ClassGraph classGraph = new ClassGraph()
                .addClassLoader(critterClassLoader)
                .enableAllInfo();
        classGraph.acceptPackages("dev.morphia.critter.sources");

        try (var scanResult = classGraph.scan()) {
            for (var classInfo : scanResult.getAllClasses()) {
                try {
                    ClassfileOutput.dump(critterClassLoader, classInfo.getName());
                } catch (Throwable ignored) {
                }
            }
        } catch (Exception ignored) {
        }

        MorphiaConfig config = MorphiaConfig.load();
        Generators generators = new Generators(config, new ReflectiveMapper(config));
        GizmoEntityModelGenerator gen = CritterGizmoGenerator.generate(Example.class, critterClassLoader, generators, false);
        try {
            entityModel = (CritterEntityModel) critterClassLoader
                    .loadClass(gen.getGeneratedType())
                    .getConstructors()[0]
                    .newInstance(generators.getMapper());
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static Object[][] methodNames(Class<?> clazz) {
        return methods(clazz).stream()
                .map(m -> new Object[] { m.getName(), m })
                .sorted(Comparator.comparing(a -> a[0].toString()))
                .toArray(Object[][]::new);
    }

    public static List<Method> methods(Class<?> clazz) {
        return Arrays.stream(clazz.getMethods())
                .filter(m -> !Modifier.isFinal(m.getModifiers()))
                .filter(m -> m.getParameterCount() == 0)
                .filter(m -> m.getDeclaringClass() == clazz)
                .collect(Collectors.toList());
    }

    /** Helper: remove list elements while predicate holds, then remove one more, return joined string. */
    static String removeWhile(List<String> list, Predicate<String> predicate) {
        List<String> removed = new ArrayList<>();
        while (!list.isEmpty() && predicate.test(list.get(0))) {
            removed.add(list.remove(0));
        }
        if (!list.isEmpty()) {
            removed.add(list.remove(0));
        }
        return String.join("\n", removed);
    }
}

File: critter/core/src/test/kotlin/dev/morphia/critter/parser/GeneratorTest.kt
package dev.morphia.critter.parser

import dev.morphia.critter.Critter.Companion.critterClassLoader
import dev.morphia.critter.parser.Generators.mapper
import dev.morphia.critter.parser.gizmo.CritterGizmoGenerator as generator
import dev.morphia.critter.parser.java.CritterParser.asmify
import dev.morphia.critter.sources.Example
import dev.morphia.mapping.codec.pojo.critter.CritterEntityModel
import io.github.classgraph.ClassGraph
import java.lang.reflect.Modifier
import org.objectweb.asm.Type

object GeneratorTest {
    var entityModel: CritterEntityModel

    init {
        val classGraph = ClassGraph().addClassLoader(critterClassLoader).enableAllInfo()
        classGraph.acceptPackages("dev.morphia.critter.sources")

        classGraph.scan().use { scanResult ->
            for (classInfo in scanResult.allClasses) {
                try {
                    val name = classInfo.name
                    critterClassLoader.dump(name)
                } catch (_: Throwable) {}
            }
        }
        val generator = generator.generate(Example::class.java)

        entityModel =
            critterClassLoader
                .loadClass(generator.generatedType)
                .constructors[0]
                .newInstance(mapper) as CritterEntityModel
    }

    fun methodNames(clazz: Class<*>): Array<Array<Any>> {
        return methods(clazz)
            .map { arrayOf(it.name, it) }
            .sortedBy { it[0].toString() }
            .toTypedArray()
    }

    fun methods(clazz: Class<*>) =
        clazz.methods
            .filterNot { method -> Modifier.isFinal(method.modifiers) }
            .filter { method -> method.parameterCount == 0 }
            .filter { method -> method.declaringClass == clazz }

    fun process(resourceName: String, entity: Type, generated: Type): String {
        var asm = asmify(critterClassLoader.getResourceAsStream(resourceName).readAllBytes())

        val lines = asm.lines()
        val imports = lines.filter { it.startsWith("import ") }
        val pkg =
            Regex("^(package )(?<pkg>.*);\$").find(lines.first())!!.groups[2]!!.value + ".__morphia"
        val body =
            asm.substringAfter("{")
                .substringBeforeLast("}")
                .substringAfter("{")
                .substringBeforeLast("}")
        var header = body.substringBefore("{")
        val methods = extractMethodDefinitions(body).map { bind(it) }.toMap(LinkedHashMap())
        return ""
        //        return "package ;"}\n" +
        //            lines.drop(1)
        //                .map { it.replace("dev/morphia/critter/sources/ExampleEntityModel",
        // "generatedType.internalName") }
        //                .joinToString("\n")
        ////            .replace()
    }

    private fun bind(methodBody: String): Pair<String, String> {
        fun extractMethodName(line: String): String {
            return line.split("\"")[1]
        }

        val lines = methodBody.lines().drop(1).dropLast(1)

        return extractMethodName(lines.first()) to lines.joinToString("\n")
    }

    private fun extractMethodDefinitions(body: String): List<String> {
        val methodBody = body.substring(body.indexOf('{')).lines().toMutableList()
        var methods = listOf<String>()
        while (methodBody.first().startsWith("{")) {
            methods += methodBody.removeWhile { it.trim() != "}" }
        }

        return methods
    }
}
